design/retryon-design.md

```diff
@@ -0,0 +1,147 @@
+      - 503
+      - 504
+    services:
+    - name: s1
+      port: 80
+```
+
+In the above example, a request would be retried on the following conditions:
+
+- An upstream connection error occurs (connection failure or timeout)
+- The response status code is 503 or 504
+
+## Detailed Design
+
+Two new fields will be added to the `v1.RetryPolicy` of the `HTTPProxy` spec:
+
+- `RetryOn []string`
+  Optional
+  Slice of [`x-envoy-retry-on`](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter#x-envoy-retry-on) conditions
+- `RetriableStatusCodes []uint32`
+  Optional
+  Slice of HTTP status codes
+
+One new field will be added to the `dag.RetryPolicy`:
+
+- `RetriableStatusCodes []uint32`
+  Slice of HTTP status codes
+
+New logic will be introduced when transforming a `v1.RetryPolicy` into a `dag.RetryPolicy`:
+
+- If `v1.RetryPolicy.RetryOn` is nil or empty, it will be set to the default value of `[]string{"5xx"}` for backwards compatibility
+- The values of `v1.RetryPolicy.RetryOn` will be joined into a string, with values separate by commas, and set as the value of `dag.RetryPolicy.RetryOn`
+- The value of `v1.RetryPolicy.RetriableStatusCodes` will be set as the value of `dag.RetryPolicy.RetriableStatusCodes`
+
+New logic will be introduced when transforming a `dag.RetryPolicy` into a Envoy v2 `route.RetryPolicy`:
+
+- The value of `dag.RetryPolicy.RetryOn` will be set as the value of `route.RetryPolicy.RetryOn`
+- The value of `dag.RetryPolicy.RetriableStatusCodes` will be set as the value of `route.RetryPolicy.RetriableStatusCodes`
+
+No further changes should be necessary to propagate these new configurable fields to Envoy.
+
+## Alternatives Considered
+
+### Exposing a single option for retrying only connection upstream errors
+
+As discussed earlier, we could expose a single option for only retrying connection upstream errors. Example:
+
+```yaml
+    retryPolicy:
+      count: 3
+      perTryTimeout: 150ms
+      retryOnConnectFailure: true
+```
+
+Arguments for this approach:
+
+- Solves the exact problem encountered and described in [Background](#background)
+
+Arguments against this approach:
+
+- Too specific to the author's problem, not general enough to apply to other use cases
+- This specific design for `HTTPProxy` is too short sighted and does not provide a path adding additional conditions in an elegant way
+
+### Explicit booleans for every valid `retry_on` value for Envoy
+
+Instead of a list of strings for `retryOn`, we could make every possible value a unique boolean field. Example:
+
+```yaml
+    retryPolicy:
+      count: 3
+      perTryTimeout: 150ms
+      retryOn:
+        5xx: false
+      	connectFailure: true
+      	gatewayError: false
+      	...
+```
+
+Arguments for this approach:
+
+- Contour can control the options that it will allow users to populate Envoy's `retry_on` field with
+- It ensures that only valid values are allowed in Envoy's `retry_on` field
```

Comment:
What happens when an invalid value is given to Envoy?  Does it do something silently which we should indeed try to avoid?

(If so, we might still prefer to check the validity in a small addition to the design, above, but it seems worth knowing and I'm relatively new to Envoy so I don't know.)

Reply:
In general, if you give an invalid value to Envoy, it will NACK the update, and you will run into #1176.

Reply:
I think that the validity of the retryOn strings should absolutely be checked against a known list, not just passed through to Envoy. It can be difficult to determine if the config has been applied correctly to Envoy, it's much better to be confident Contour has passed a correct config.

Reply:
+1 to what @youngnick said, Contour should always strive to produce clean config before passing off to envoy.

Reply:
Agree about the validation before sending to Envoy. I'll update the design proposal to solve this with `kubebuilder` validation markers. I still think it makes sense to use a `[]string` for `RetryOn` as using a unique boolean field for each possible value seems too verbose.